server/core/region_tree.go

```diff
@@ -161,34 +162,85 @@ func (t *regionTree) find(region *metapb.Region) *regionItem {
 
 // scanRage scans from the first region containing or behind the start key
 // until f return false
-func (t *regionTree) scanRange(startKey []byte, f func(*metapb.Region) bool) {
-	region := &metapb.Region{StartKey: startKey}
+func (t *regionTree) scanRange(startKey []byte, f func(*RegionInfo) bool) {
+	region := &RegionInfo{meta: &metapb.Region{StartKey: startKey}}
 	// find if there is a region with key range [s, d), s < startKey < d
 	startItem := t.find(region)
 	if startItem == nil {
-		startItem = &regionItem{region: &metapb.Region{StartKey: startKey}}
+		startItem = &regionItem{region: &RegionInfo{meta: &metapb.Region{StartKey: startKey}}}
 	}
 	t.tree.AscendGreaterOrEqual(startItem, func(item btree.Item) bool {
 		return f(item.(*regionItem).region)
 	})
 }
 
-func (t *regionTree) getAdjacentRegions(region *metapb.Region) (*regionItem, *regionItem) {
-	item := &regionItem{region: &metapb.Region{StartKey: region.StartKey}}
+func (t *regionTree) getAdjacentRegions(region *RegionInfo) (*regionItem, *regionItem) {
+	item := &regionItem{region: &RegionInfo{meta: &metapb.Region{StartKey: region.GetStartKey()}}}
 	var prev, next *regionItem
 	t.tree.AscendGreaterOrEqual(item, func(i btree.Item) bool {
-		if bytes.Equal(item.region.StartKey, i.(*regionItem).region.StartKey) {
+		if bytes.Equal(item.region.GetStartKey(), i.(*regionItem).region.GetStartKey()) {
 			return true
 		}
 		next = i.(*regionItem)
 		return false
 	})
 	t.tree.DescendLessOrEqual(item, func(i btree.Item) bool {
-		if bytes.Equal(item.region.StartKey, i.(*regionItem).region.StartKey) {
+		if bytes.Equal(item.region.GetStartKey(), i.(*regionItem).region.GetStartKey()) {
 			return true
 		}
 		prev = i.(*regionItem)
 		return false
 	})
 	return prev, next
 }
+
+func (t *regionTree) RandomRegion(startKey, endKey []byte) *RegionInfo {
+	if t.length() == 0 {
+		return nil
+	}
+
+	// The tree only contains one region.
+	if t.length() == 1 {
+		return t.tree.Min().(*regionItem).region
+	}
+
+	var startIndex, endIndex, index int
+	var startRegion, endRegion *RegionInfo
+
+	if len(startKey) != 0 {
+		startRegion, startIndex = t.getWithIndex(&regionItem{region: &RegionInfo{meta: &metapb.Region{StartKey: startKey}}})
+	} else {
+		startRegion, startIndex = t.getWithIndex(t.tree.Min())
+	}
+
+	if len(endKey) != 0 {
+		endRegion, endIndex = t.getWithIndex(&regionItem{region: &RegionInfo{meta: &metapb.Region{StartKey: endKey}}})
+	} else {
+		_, endIndex = t.getWithIndex(t.tree.Max())
+		endRegion = nil
+		endIndex++
+	}
+
+	if endIndex == startIndex {
+		if endRegion == nil {
+			return t.tree.GetAt(startIndex - 1).(*regionItem).region
+		}
```

Comment:
What does this mean?
What's more, `startIndex` may be 0, then `t.tree.GetAt(-1)` will be `nil`. This may panic.

Reply:
It won't panic because only if `startIndex==endindex`, it will enter this block. `if t.length() == 1` will ensure that the tree has more than one region.

Reply:
The index of the first item is 0.